File: haskell/tests/BinarySearchTest.hs
module BinarySearchTest (tests) where

import Test.HUnit
import BinarySearch

-- Test cases
tests :: Test
tests = TestList [
    "test_binary_search" ~: TestList [
        "target in middle" ~: assertEqual "Should find target in middle" (Just 2) (binarySearch [1,2,3,4,5] 3),
        "target at start" ~: assertEqual "Should find target at start" (Just 0) (binarySearch [1,2,3,4,5] 1),
        "target at end" ~: assertEqual "Should find target at end" (Just 4) (binarySearch [1,2,3,4,5] 5),
        "target not present" ~: assertEqual "Should return Nothing when target not found" Nothing (binarySearch [1,2,3,4,5] 6),
        "empty list" ~: assertEqual "Should handle empty list" Nothing (binarySearch [] 1),
        "single element" ~: assertEqual "Should handle single element" (Just 0) (binarySearch [1] 1),
        "negative numbers" ~: assertEqual "Should handle negative numbers" (Just 2) (binarySearch [-3,-2,-1,0,1] -1),
        "duplicate elements" ~: assertEqual "Should find first occurrence" (Just 1) (binarySearch [1,2,2,2,3] 2)
    ]
]

main :: IO Counts
main = runTestTT tests

File: go/tests/binary_search_test.go
package main

import "testing"

func TestBinarySearch(t *testing.T) {
	tests := []struct {
		name     string
		arr      []int
		target   int
		expected int
	}{
		{"target in middle", []int{1, 2, 3, 4, 5}, 3, 2},
		{"target at start", []int{1, 2, 3, 4, 5}, 1, 0},
		{"target at end", []int{1, 2, 3, 4, 5}, 5, 4},
		{"target not present", []int{1, 2, 3, 4, 5}, 6, -1},
		{"empty array", []int{}, 1, -1},
		{"single element", []int{1}, 1, 0},
		{"negative numbers", []int{-5, -3, -1, 0, 2}, -3, 1},
		{"duplicate elements", []int{1, 2, 2, 2, 3}, 2, 1},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := BinarySearch(tt.arr, tt.target)
			if result != tt.expected {
				t.Errorf("BinarySearch(%v, %d) = %d; want %d",
					tt.arr, tt.target, result, tt.expected)
			}
		})
	}
}
